include/dlaf/communication/executor.h

```diff
@@ -0,0 +1,162 @@
+auto make_mpi_tuple(Tuple t, MPI_Request* req_ptr) {
+  return hpx::tuple_cat(std::move(t), hpx::make_tuple(req_ptr));
+}
+
+// Wraps the invocation of `F` such that the void and non-void cases are handled without code duplication.
+template <class R>
+struct invoke_fused_wrapper {
+  R val;
+  template <class F, class TupleArgs>
+  invoke_fused_wrapper(F&& f, TupleArgs&& ts)
+      : val(hpx::invoke_fused(std::forward<F>(f), std::forward<TupleArgs>(ts))) {}
+  R async_return() {
+    return std::move(val);
+  }
+  R dataflow_return() {
+    return std::move(val);
+  }
+};
+
+template <>
+struct invoke_fused_wrapper<void> {
+  template <class F, class TupleArgs>
+  invoke_fused_wrapper(F&& f, TupleArgs&& ts) {
+    hpx::invoke_fused(std::forward<F>(f), std::forward<TupleArgs>(ts));
+  }
+  void async_return() {}
+  auto dataflow_return() {
+    return hpx::util::unused;
+  }
+};
+
+}
+
+class Executor {
+  hpx::execution::parallel_executor ex_;
+  MPIMech mech_;
+
+public:
+  // Notes:
+  //   - MPI event polling has to be enabled for `MPIMech::Polling`.
+  Executor(const std::string& pool, MPIMech mech = MPIMech::Yielding)
+      : ex_(&hpx::resource::get_thread_pool(pool)), mech_(mech) {}
+
+  bool operator==(const Executor& rhs) const noexcept {
+    return ex_ == rhs.ex_;
+  }
+
+  bool operator!=(const Executor& rhs) const noexcept {
+    return !(*this == rhs);
+  }
+
+  const Executor& context() const noexcept {
+    return *this;
+  }
+
+  template <typename F, typename... Ts>
+  auto async_execute(F&& f, Ts&&... ts) noexcept {
+    auto fn = [mech = mech_, f = std::forward<F>(f),
+               args = hpx::make_tuple(std::forward<Ts>(ts)...)]() mutable {
+      MPI_Request req;
+      auto all_args = detail::make_mpi_tuple(std::move(args), &req);
+      using result_t = decltype(hpx::util::invoke_fused(f, all_args));
+      detail::invoke_fused_wrapper<result_t> wrapper(std::move(f), std::move(all_args));
+      detail::handle_request(mech, req);
+      return wrapper.async_return();
+    };
+    return hpx::async(ex_, std::move(fn));
+  }
+
+  template <class Frame, class F, class TupleArgs>
+  void dataflow_finalize(Frame&& frame, F&& f, TupleArgs&& args) {
+    // Ensure the dataflow frame stays alive long enough.
+    using FramePtr =
+        hpx::intrusive_ptr<typename std::remove_pointer<typename std::decay<Frame>::type>::type>;
+    FramePtr frame_p(std::forward<Frame>(frame));
+    auto fn = [mech = mech_, frame_p = std::move(frame_p), f = std::forward<F>(f),
+               args = std::forward<TupleArgs>(args)]() mutable {
+      MPI_Request req;
+      auto all_args = detail::make_mpi_tuple(std::move(args), &req);
+      using result_t = decltype(hpx::util::invoke_fused(f, all_args));
+      detail::invoke_fused_wrapper<result_t> wrapper(std::move(f), std::move(all_args));
+      detail::handle_request(mech, req);
+      frame_p->set_data(wrapper.dataflow_return());
+    };
+    hpx::async(ex_, std::move(fn));
```

Comment:
Have I understood it correctly that you currently do the polling on the default pool? Also, does the pipeline wrapper have to ensure that there is only ever a single MPI request in flight, or that they are only submitted in order? Is the wrapper in args, and in that case the continuation is triggered only once `fn` has finished?

For the polling case, it would be slightly preferable to submit the MPI operation when entering `dataflow_finalize` and attach a continuation to the future from `hpx::mpi::experimental::get_future` which calls `set_data` which would avoid a probable yield when calling `get` on the future. However, this is just a potential optimization.

I don't see anything obvious right now that would cause it to deadlock but I'll have a closer look still. When you say it hangs after `hpx::finalize`, does that mean you see `test_cholesky` actually successfully run all the tests, but it hangs after that?

Reply:
> Have I understood it correctly that you currently do the polling on the default pool? 

Polling can be done both on the default and the mpi pools. Unit tests, do it on the default pool. Miniapps do it on the mpi pool currently, but I will implement a command line option allowing to switch between the default and mpi pools.

> Also, does the pipeline wrapper have to ensure that there is only ever a single MPI request in flight, or that they are only submitted in order?

Only that they are submitted in order on the same communicator.

> Is the wrapper in args, and in that case the continuation is triggered only once fn has finished?

Yes, the pipeline wrapper is in `args`. The underlying promise is set when `fn` finishes using RAII (if that is what you are asking).

> For the polling case, it would be slightly preferable to submit the MPI operation when entering dataflow_finalize and attach a continuation to the future from hpx::mpi::experimental::get_future which calls set_data which would avoid a probable yield when calling get on the future. However, this is just a potential optimization.

I agree that this is the case for the non-blocking MPI approach (which is what we are left with in this PR). However, for blocking MPI, the call can't happen in `dataflow_finalize()` and has to be using the executor `ex_`, because it will block the underlying OS thread on which `dataflow_finalize()` is scheduled.

> I don't see anything obvious right now that would cause it to deadlock but I'll have a closer look still. When you say it hangs after hpx::finalize, does that mean you see test_cholesky actually successfully run all the tests, but it hangs after that?

Thanks for helping me on this one. The issue was that polling was disabled before all outstanding non-blocking communications were done.